mayastor/src/bdev/nexus/nexus_metadata.rs

```diff
@@ -0,0 +1,620 @@
+                return Err(MetaDataError::PartitionSizeExceeded {});
+            }
+
+            let mut buf =
+                DmaBuf::new((blocks * block_size) as usize, bdev.alignment())
+                    .context(ObjectAlloc {})?;
+            let mut writer = Cursor::new(buf.as_mut_slice());
+
+            serialize_into(&mut writer, config).context(SerializeError {})?;
+            let checksum = crc32::checksum_ieee(buf.as_slice());
+
+            self.write_at((header.self_lba + start) * block_size, &buf)
+                .await
+                .context(ObjectWrite {})?;
+
+            revision += 1;
+
+            index.push(MetaDataIndexEntry {
+                revision,
+                timestamp,
+                data_checksum: checksum,
+                data_start: start,
+                data_end: next - 1,
+            });
+
+            start = next;
+        }
+
+        Ok(index)
+    }
+
+    // Determines if the data defined by the index is currently fragmented.
+    // Returns an error if the index is inconsistent.
+    fn fragmented(
+        mut start: u64,
+        index: &[MetaDataIndexEntry],
+    ) -> Result<bool, MetaDataError> {
+        for entry in index {
+            if entry.data_start < start {
+                return Err(MetaDataError::IndexInconsistent {});
+            }
+            if entry.data_start > start {
+                return Ok(true);
+            }
+            start = entry.data_end + 1;
+        }
+        Ok(false)
+    }
+
+    // Defragment the data defined by the index, and update the index in situ.
+    async fn compact(
+        &mut self,
+        metadata: &mut NexusMetaData,
+    ) -> Result<(), MetaDataError> {
+        let (bdev, _desc) = self.get_dev()?;
+        let block_size = bdev.block_len() as u64;
+        let alignment = bdev.alignment();
+
+        let self_lba = metadata.header.self_lba;
+        let mut start = metadata.header.data_start;
+
+        for entry in &mut metadata.index {
+            if entry.data_start > start {
+                let blocks = entry.data_end - entry.data_start;
+                let mut buf = DmaBuf::new(
+                    ((blocks + 1) * block_size) as usize,
+                    alignment,
+                )
+                .context(ObjectAlloc {})?;
+                self.read_at(
+                    (self_lba + entry.data_start) * block_size,
+                    &mut buf,
+                )
+                .await
+                .context(ObjectRead {})?;
+                self.write_at((self_lba + start) * block_size, &buf)
+                    .await
+                    .context(ObjectWrite {})?;
+                entry.data_start = start;
+                entry.data_end = start + blocks;
+            }
+            start = entry.data_end + 1;
+        }
+
+        Ok(())
+    }
+
+    // Update checksums and write out MetaData header + index to disk.
+    pub async fn sync_metadata(
```

Comment:
This syncs the header + index, not all the metadata right? problably we want to be explicitit about that?